watchman.c

```diff
@@ -179,6 +179,32 @@ int watchman_pclose(struct watchman_error *error, struct watchman_popen *popen)
     }
 }
 
+/**
+ * Call select() on the given fd until it's ready to read from or
+ * until the timeout expires. Updates timeout to indicate the
+ * remaining time. Returns 1 if the socket is readable, 0 if the
+ * socket is not readable, and -1 on error.
+ */
+static int block_on_read(int fd, struct timeval *timeout)
+{
+    struct timeval now, start, elapsed;
+    fd_set fds;
+    int ret=0;
+
+    FD_ZERO(&fds);
+    FD_SET(fd, &fds);
+
+    gettimeofday(&start, NULL);
+
+    ret = select(fd + 1, &fds, NULL, NULL, timeout);
```

Comment:
We need to make a copy of timeout here:
"On Linux, select() modifies timeout to reflect the amount of time not slept; most other implementations do not do this.  (POSIX.1-2001 permits either behavior.)"